### ISSUE TITLE
Creating a package while specifying --service-account-name flag fails

### ISSUE THREAD

**Issue by kacole2** [vmware-tanzu/tanzu-framework#1325]
**Bug description**
Running an installation when specifying the flag `--service-account-name` will consistently fail, even when using the default service account. 

I am running TCE behind the VMware corporate firewall on a vSphere cluster. It's impossible to install the contour package since it pulls from DockerHub and rate limiting is reached. I am trying to find a solution where I can install a tanzu package using my Docker Hub credentials. I have created a Kubernetes secret with my Docker Hub credentials and patched the default service account with `imagePullSecrets`. I have verified a simple nginx deployment works. Specifying a `--service-account-name default` on the `tanzu package install` fails even though I believe that is the service account being used when the flag is not specified.

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ X] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ X] Installation
- [ X] Plugin
- [ ] Security
- [ ] Test and Release
- [ X] User Experience

**Expected behavior**
Tanzu packages should be able to use the `--service-account-name` flag. Furthermore, a service account that has `imagePullSecrets` assigned to pull images from DockerHub.

**Steps to reproduce the bug**
The following command should install contour in the default namespace using the default service account
```
tanzu package install contour --package-name contour.community.tanzu.vmware.com --version 1.18.1 --service-account-name default
```

The error given:
```
Error: package install deletion failed: kapp: Error: Getting app: configmaps "contour2-ctrl" is forbidden:
  User "system:serviceaccount:default:default" cannot get resource "configmaps" in API group "" in the namespace "default"
```

A side effect of this is that you cannot delete the package at all.  The status is delete failed
```
kendrickc@kendrickc-a01 ~ % tanzu package installed list
- Retrieving installed packages...
  NAME      PACKAGE-NAME                        PACKAGE-VERSION  STATUS
  contour   contour.community.tanzu.vmware.com  1.18.1           Delete failed: Error (see .status.usefulErrorMessage for details)
  contour2  contour.community.tanzu.vmware.com  1.18.1           Delete failed: Error (see .status.usefulErrorMessage for details)
  contour3  contour.community.tanzu.vmware.com  1.18.1           Reconcile failed: Error (see .status.usefulErrorMessage for details)
```

```
kendrickc@kendrickc-a01 ~ % tanzu package installed get contour
/ Retrieving installation details for contour...
NAME:                    contour
PACKAGE-NAME:            contour.community.tanzu.vmware.com
PACKAGE-VERSION:         1.18.1
STATUS:                  Delete failed: Error (see .status.usefulErrorMessage for details)
CONDITIONS:              [{DeleteFailed True  Error (see .status.usefulErrorMessage for details)}]
USEFUL-ERROR-MESSAGE:    kapp: Error: Getting app: configmaps "contour-ctrl" is forbidden:
  User "system:serviceaccount:default:default" cannot get resource "configmaps" in API group "" in the namespace "default"
```

**Version** (include the SHA if the version is not obvious)
```
kendrickc@kendrickc-a01 ~ % tanzu version
version: v0.2.1
buildDate: 2021-09-29
sha: ceaa474
```

**Environment where the bug was observed (cloud, OS, etc)**
TCE running on Mac OS and environment is vSphere 7 running inside the VMware network

**Relevant Debug Output (Logs, manifests, etc)**

**Comment by maralavi** [comment]
Hi @kacole2 .  Actually if no service account is provided, the "default" is not used, CLI creates one using package install and namespace name.

The output below seems unrelated to package install, and relates to package installed delete. Could I ask you please share the error message you got for package install? You can get from ```"kubectl get pkgi contour -o yaml"```

```
Error: package install deletion failed: kapp: Error: Getting app: configmaps "contour2-ctrl" is forbidden:
  User "system:serviceaccount:default:default" cannot get resource "configmaps" in API group "" in the namespace "default"
```

**Comment by kacole2** [comment]
@maralavi here is some output:

```
kendrickc@kendrickc-a01 ~ % tanzu package installed list
/ Retrieving installed packages...
  NAME          PACKAGE-NAME                             PACKAGE-VERSION  STATUS
  cert-manager  cert-manager.community.tanzu.vmware.com  1.5.3            Reconcile succeeded
  contour       contour.community.tanzu.vmware.com       1.18.1           Delete failed: Error (see .status.usefulErrorMessage for details)
  contour2      contour.community.tanzu.vmware.com       1.18.1           Delete failed: Error (see .status.usefulErrorMessage for details)
  contour3      contour.community.tanzu.vmware.com       1.18.1           Reconcile failed: Error (see .status.usefulErrorMessage for details)
kendrickc@kendrickc-a01 ~ % tanzu package installed get contour3
/ Retrieving installation details for contour3...
NAME:                    contour3
PACKAGE-NAME:            contour.community.tanzu.vmware.com
PACKAGE-VERSION:         1.18.1
STATUS:                  Reconcile failed: Error (see .status.usefulErrorMessage for details)
CONDITIONS:              [{ReconcileFailed True  Error (see .status.usefulErrorMessage for details)}]
USEFUL-ERROR-MESSAGE:    kapp: Error: Creating app: configmaps is forbidden:
  User "system:serviceaccount:default:dhcreds" cannot create resource "configmaps" in API group "" in the namespace "default"
```

```
kendrickc@kendrickc-a01 ~ % kubectl get pkgi contour3 -o yaml
apiVersion: packaging.carvel.dev/v1alpha1
kind: PackageInstall
metadata:
  creationTimestamp: "2021-12-07T22:44:20Z"
  finalizers:
  - finalizers.packageinstall.packaging.carvel.dev/delete
  generation: 1
  name: contour3
  namespace: default
  resourceVersion: "446717"
  uid: 9050f9d7-c7f3-4ac5-a4aa-f2d81c59ef0b
spec:
  packageRef:
    refName: contour.community.tanzu.vmware.com
    versionSelection:
      constraints: 1.18.1
      prereleases: {}
  serviceAccountName: dhcreds
status:
  conditions:
  - message: Error (see .status.usefulErrorMessage for details)
    status: "True"
    type: ReconcileFailed
  friendlyDescription: 'Reconcile failed: Error (see .status.usefulErrorMessage for
    details)'
  observedGeneration: 1
  usefulErrorMessage: |-
    kapp: Error: Creating app: configmaps is forbidden:
      User "system:serviceaccount:default:dhcreds" cannot create resource "configmaps" in API group "" in the namespace "default"
  version: 1.18.1
```

**Comment by maralavi** [comment]
Thanks @kacole2 for sharing the above outputs. Could you please verify that you have created clusterrolebinding as well? It should reference to clustertole and service account you created. I suspect the problem is with that part.

**Comment by kacole2** [comment]
@maralavi that's likely the cause. Nothing in the documentation specified that a clusterrolebinding was necessary. I've tried this and I'm still getting the same error. Any pointers?

```
kubectl apply -f - <<EOF
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: all-access
rules:
- apiGroups: [""]
  resources: ["*"]
  verbs: ["*"]
EOF

kubectl apply -f - <<EOF
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: all-access-global
subjects:
- kind: User
  name: default
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: all-access
  apiGroup: rbac.authorization.k8s.io
EOF
```

```
kendrickc@kendrickc-a01 ~ % tanzu package install contour4 --package-name contour.community.tanzu.vmware.com --version 1.18.1 --service-account-name default

| Installing package 'contour.community.tanzu.vmware.com'
/ Getting namespace 'default'
\ Getting package metadata for 'contour.community.tanzu.vmware.com'
\ Creating package resource
| Package install status: Reconciling
| Package install status: ReconcileFailed

Please consider using 'tanzu package installed update' to update the installed package with correct settings


Error: package reconciliation failed: kapp: Error: Creating app: configmaps is forbidden:
  User "system:serviceaccount:default:default" cannot create resource "configmaps" in API group "" in the namespace "default"
```

**Comment by maralavi** [comment]
I believe you need to add something like the following to the clusterrolebinding for the service account (it is of kind ServiceAccount):
```
subjects:
- kind: ServiceAccount
  name: <NAME>
  namespace: <NAMESPACE>
```

@kacole2 True. the CLI creates service account, cluster role and clusterrolebinding itself if the user did not provided --service-account-name, but if the user provides it, it is assumed that they've configured those 3 properly, that's why its not documented in the CLI related docs:(

**Comment by kacole2** [comment]
I'm not sure why this is still not working. The ServiceAccount of default has the DockerHub credentials secret and I've verified that is working with a simple deployment.

```
kendrickc@kendrickc-a01 ~ % tanzu package installed get contour5
- Retrieving installation details for contour5...
NAME:                    contour5
PACKAGE-NAME:            contour.community.tanzu.vmware.com
PACKAGE-VERSION:         1.18.1
STATUS:                  Reconcile failed: Error (see .status.usefulErrorMessage for details)
CONDITIONS:              [{ReconcileFailed True  Error (see .status.usefulErrorMessage for details)}]
USEFUL-ERROR-MESSAGE:    kapp: Error: waiting on reconcile deployment/contour (apps/v1) namespace: projectcontour:
  Finished unsuccessfully (Deployment is not progressing: ProgressDeadlineExceeded (message: ReplicaSet "contour-5c6cc944c4" has timed out progressing.))
kendrickc@kendrickc-a01 ~ % kubectl get pods -A
NAMESPACE        NAME                                                     READY   STATUS              RESTARTS   AGE
avi-system       ako-0                                                    1/1     Running             0          6d22h
cert-manager     cert-manager-868cbf47df-5rgsh                            1/1     Running             0          5d19h
cert-manager     cert-manager-cainjector-5549cd8d64-49bfc                 1/1     Running             0          5d19h
cert-manager     cert-manager-webhook-77f5768fdf-lt7k8                    1/1     Running             0          5d19h
default          nginx-6799fc88d8-b4kpm                                   1/1     Running             0          6d20h
kube-system      antrea-agent-hbc7f                                       2/2     Running             0          6d22h
kube-system      antrea-agent-jpwrb                                       2/2     Running             0          6d22h
kube-system      antrea-controller-6cc9dd86c-pgcs6                        1/1     Running             0          6d22h
kube-system      coredns-8dcb5c56b-nfn67                                  1/1     Running             0          6d22h
kube-system      coredns-8dcb5c56b-pn222                                  1/1     Running             0          6d22h
kube-system      etcd-tcevmw01-control-plane-h8wnv                        1/1     Running             0          6d22h
kube-system      kube-apiserver-tcevmw01-control-plane-h8wnv              1/1     Running             0          6d22h
kube-system      kube-controller-manager-tcevmw01-control-plane-h8wnv     1/1     Running             0          6d22h
kube-system      kube-proxy-5wlcq                                         1/1     Running             0          6d22h
kube-system      kube-proxy-7z9qt                                         1/1     Running             0          6d22h
kube-system      kube-scheduler-tcevmw01-control-plane-h8wnv              1/1     Running             0          6d22h
kube-system      metrics-server-7c95d7d465-ndvcw                          1/1     Running             0          6d22h
kube-system      vsphere-cloud-controller-manager-9d76l                   1/1     Running             0          6d22h
kube-system      vsphere-csi-controller-7dc87cb98b-7bd99                  6/6     Running             0          6d22h
kube-system      vsphere-csi-node-64qfj                                   3/3     Running             0          6d22h
kube-system      vsphere-csi-node-gvrk6                                   3/3     Running             0          6d22h
projectcontour   contour-5c6cc944c4-c9snt                                 0/1     ContainerCreating   0          19h
projectcontour   contour-5c6cc944c4-t97jl                                 0/1     ContainerCreating   0          19h
projectcontour   contour-certgen-v1.18.1-ph5bs                            0/1     ImagePullBackOff    0          19h
projectcontour   envoy-jxrfx                                              0/2     Init:0/1            0          19h
tkg-system       kapp-controller-54f6797dbf-zqlxp                         1/1     Running             0          6d22h
tkg-system       tanzu-capabilities-controller-manager-6ff97656b8-r2qdd   1/1     Running             0          6d22h
kendrickc@kendrickc-a01 ~ %
```

**Comment by kacole2** [comment]
Some more debugging

```
kendrickc@kendrickc-a01 ~ % kubectl describe pod contour-5c6cc944c4-t97jl -n projectcontour
Name:           contour-5c6cc944c4-t97jl
Namespace:      projectcontour
Priority:       0
Node:           tcevmw01-md-0-7dc987754c-pk4sv/10.202.121.59
Start Time:     Mon, 13 Dec 2021 17:10:25 -0500
Labels:         app=contour
                kapp.k14s.io/app=1639433417137954950
                kapp.k14s.io/association=v1.0285c703a2b28b8f555fa675e048a135
                pod-template-hash=5c6cc944c4
Annotations:    prometheus.io/port: 8000
                prometheus.io/scrape: true
Status:         Pending
IP:
IPs:            <none>
Controlled By:  ReplicaSet/contour-5c6cc944c4
Containers:
  contour:
    Container ID:
    Image:         index.docker.io/projectcontour/contour@sha256:31c376f80f5b80f2ac5558c1c50711438796b22b742285789ea21cd2bba244e2
    Image ID:
    Ports:         8001/TCP, 8000/TCP, 6060/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Command:
      contour
    Args:
      serve
      --incluster
      --xds-address=0.0.0.0
      --xds-port=8001
      --contour-cafile=/certs/ca.crt
      --contour-cert-file=/certs/tls.crt
      --contour-key-file=/certs/tls.key
      --config-path=/config/contour.yaml
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Liveness:       http-get http://:8000/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Readiness:      tcp-socket :8001 delay=15s timeout=1s period=10s #success=1 #failure=3
    Environment:
      CONTOUR_NAMESPACE:  projectcontour (v1:metadata.namespace)
      POD_NAME:           contour-5c6cc944c4-t97jl (v1:metadata.name)
    Mounts:
      /certs from contourcert (ro)
      /config from contour-config (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-xh6h8 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  contourcert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  contourcert
    Optional:    false
  contour-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      contour
    Optional:  false
  kube-api-access-xh6h8:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                    From     Message
  ----     ------       ----                   ----     -------
  Warning  FailedMount  49m (x115 over 19h)    kubelet  Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[kube-api-access-xh6h8 contourcert contour-config]: timed out waiting for the condition
  Warning  FailedMount  13m (x580 over 19h)    kubelet  MountVolume.SetUp failed for volume "contourcert" : secret "contourcert" not found
  Warning  FailedMount  8m42s (x305 over 19h)  kubelet  Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contourcert contour-config kube-api-access-xh6h8]: timed out waiting for the condition
  Warning  FailedMount  4m10s (x96 over 19h)   kubelet  Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contour-config kube-api-access-xh6h8 contourcert]: timed out waiting for the condition
```

same thing on another pod.
```
Events:
  Type     Reason       Age                    From     Message
  ----     ------       ----                   ----     -------
  Warning  FailedMount  20m (x317 over 19h)    kubelet  Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contourcert contour-config kube-api-access-ftxs7]: timed out waiting for the condition
  Warning  FailedMount  15m (x94 over 19h)     kubelet  Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contour-config kube-api-access-ftxs7 contourcert]: timed out waiting for the condition
  Warning  FailedMount  4m46s (x585 over 19h)  kubelet  MountVolume.SetUp failed for volume "contourcert" : secret "contourcert" not found
```

```
Events:
  Type     Reason       Age                   From     Message
  ----     ------       ----                  ----     -------
  Warning  FailedMount  60m (x97 over 19h)    kubelet  Unable to attach or mount volumes: unmounted volumes=[envoycert], unattached volumes=[envoy-admin envoy-config envoycert]: timed out waiting for the condition
  Warning  FailedMount  26m (x311 over 19h)   kubelet  Unable to attach or mount volumes: unmounted volumes=[envoycert], unattached volumes=[envoy-config envoycert envoy-admin]: timed out waiting for the condition
  Warning  FailedMount  6m4s (x103 over 19h)  kubelet  Unable to attach or mount volumes: unmounted volumes=[envoycert], unattached volumes=[envoycert envoy-admin envoy-config]: timed out waiting for the condition
  Warning  FailedMount  108s (x587 over 19h)  kubelet  MountVolume.SetUp failed for volume "envoycert" : secret "envoycert" not found
```

**Comment by maralavi** [comment]
@kacole2  Your error message seems to have changed, right? Could you please provide output for the following? ```kubectl get pkgi contour5 -o yaml```? 
Also please makes sure you only have one package install for contour. If you get more than one "contour" entry from ```kubectl get pkgi```, please delete those by running ```tanzu package installed delete <PACKAGE_INSTALL-NAME> -y```

**Comment by kacole2** [comment]
Here is the output. The process starts now but the pods never finish installation because it's looking for certs. 

```
kendrickc@kendrickc-a01 ~ % kubectl get pkgi contour5 -o yaml
apiVersion: packaging.carvel.dev/v1alpha1
kind: PackageInstall
metadata:
  creationTimestamp: "2021-12-13T22:10:09Z"
  finalizers:
  - finalizers.packageinstall.packaging.carvel.dev/delete
  generation: 1
  name: contour5
  namespace: default
  resourceVersion: "2274566"
  uid: 0a57239e-db85-40cf-a019-2b73e0bd97be
spec:
  packageRef:
    refName: contour.community.tanzu.vmware.com
    versionSelection:
      constraints: 1.18.1
      prereleases: {}
  serviceAccountName: default
status:
  conditions:
  - message: Error (see .status.usefulErrorMessage for details)
    status: "True"
    type: ReconcileFailed
  friendlyDescription: 'Reconcile failed: Error (see .status.usefulErrorMessage for
    details)'
  observedGeneration: 1
  usefulErrorMessage: |-
    kapp: Error: waiting on reconcile deployment/contour (apps/v1) namespace: projectcontour:
      Finished unsuccessfully (Deployment is not progressing: ProgressDeadlineExceeded (message: ReplicaSet "contour-5c6cc944c4" has timed out progressing.))
  version: 1.18.1
```

I can try deleting the cluster and starting from scratch. There are currently failed deployments that will not delete because they didn't reconcile. doing a `tanzu package installed delete contour` will fail because the `--service-account-name` flag failed during deployment.

**Comment by maralavi** [comment]
@kacole2 I assume you have cert manager installed properly right? hmm, why "package installed delete" is failing? Do you have output for that? Did you give it enough time to finish deletion?

**Comment by kacole2** [comment]
cert-manager installed without issues. I just tried installing this on a brand new cluster and I'm still getting the same issues.

```
  Normal   Scheduled    6m52s                 default-scheduler  Successfully assigned projectcontour/contour-6f9b79cd56-8ldnm to tcevmw02-md-0-6c7bdf4ff4-sh5wd
  Warning  FailedMount  4m49s                 kubelet            Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contour-config kube-api-access-tlxxp contourcert]: timed out waiting for the condition
  Warning  FailedMount  40s (x11 over 6m52s)  kubelet            MountVolume.SetUp failed for volume "contourcert" : secret "contourcert" not found
  Warning  FailedMount  18s (x2 over 2m35s)   kubelet            Unable to attach or mount volumes: unmounted volumes=[contourcert], unattached volumes=[contourcert contour-config kube-api-access-tlxxp]: timed out waiting for the condition
```

There is some type of permission issue using the `--service-account-name` flag that is not able to find the secrets. Do I need to pre-create the namespace `projectcontour` and then create a service account and secret in that namespace before installing?

**Comment by blc1996** [comment]
Hi @kacole2 , could you try editing the Kapp's placeholder secret for Contour PKGI as listed in https://carvel.dev/kapp-controller/docs/latest/private-registry-auth/#bringing-it-all-together.
It should be a secret ending with `fetch-0`.

**Comment by kacole2** [comment]
I'm confused. Where is `the Kapp's placeholder secret for Contour PKGI`? Is this installed somewhere? Is it already downloaded? Am I supposed to edit the tce-repo packagerepository? Do I have to manually create the `projectcontour` namespace prior to running? The documentation isn't very clear on a concise way to accomplish what I'm trying to do.

**Comment by maralavi** [comment]
The original issue with --service-account-name works as expected )just some setups were needed to configure the user-created service account).

@kacole2 , could we close this issue and if needed you kindly create a new issue for the issue with docker hub registry where you were getting "too many requests" error?

**Comment by kacole2** [comment]
I'll work on creating a new issue. Thanks.
